services/streammanager/kafka/client/producer.go

```diff
@@ -140,28 +138,45 @@ func headers(msg Message) (headers []kafka.Header) {
 }
 
 func isErrTemporary(err error) bool {
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
 	isTransientNetworkError := errors.Is(err, io.ErrUnexpectedEOF) ||
 		errors.Is(err, syscall.ECONNREFUSED) ||
 		errors.Is(err, syscall.ECONNRESET) ||
 		errors.Is(err, syscall.EPIPE)
 	if isTransientNetworkError {
 		return true
 	}
+	var tempError interface{ Temporary() bool }
 	if errors.As(err, &tempError) {
 		return tempError.Temporary()
 	}
+	var timeoutError interface{ Timeout() bool }
+	if errors.As(err, &timeoutError) {
+		return timeoutError.Timeout()
+	}
 	return false
 }
 
 func IsProducerErrTemporary(err error) bool {
-	if we, ok := err.(kafka.WriteErrors); ok {
-		for _, err := range we {
-			// if at least one was temporary then we treat the whole batch as such
-			if isErrTemporary(err) {
-				return true
+	f := func(err error) bool {
+		if err == nil {
+			return false
+		}
+		if we, ok := err.(kafka.WriteErrors); ok {
+			for _, err := range we {
+				// if at least one was temporary then we treat the whole batch as such
+				if isErrTemporary(err) {
+					return true
+				}
 			}
+			return false
 		}
-		return false
+		return isErrTemporary(err)
+	}
+	if ue := errors.Unwrap(err); ue != nil {
+		return f(err) || IsProducerErrTemporary(ue)
```

Comment:
couldn't we spare this if we simply converted 
```
if we, ok := err.(kafka.WriteErrors); ok {
```
to
```
var we kafka.WriteErrors
if errors.As(err, &we) {
```
?
If we make sure all checks inside `f` properly unwrap errors, we wouldn't need to unwrap them manually here and perform the recursive operation afterwards

Reply:
True but it is slower given that it uses reflection on top of `Unwrap`. It's still looping (without recursion) so it works. In any case for the sake of simplicity I will change it. 

Btw, another trick is that I was doing:
```go
if ue := errors.Unwrap(err); ue != nil {
	return f(err) || IsProducerErrTemporary(ue)
}
return f(err)
```

So recursion was not used unless the parent error was not temporary already. Might even be nicer this way to avoid calling `Unwrap` unnecessarily:

```go
if f(err) {
    return true
} else if ue := errors.Unwrap(err); ue != nil {
	return IsProducerErrTemporary(ue)
}
return false
```

```
cpu: 12th Gen Intel(R) Core(TM) i9-12900K
BenchmarkTempErrors
BenchmarkTempErrors/recursion
BenchmarkTempErrors/recursion-24         	 4419303	       266.6 ns/op
BenchmarkTempErrors/errors.As
BenchmarkTempErrors/errors.As-24         	 3531273	       350.4 ns/op
PASS
```

In any case, thanks for the comment! :+1: